crates/accelerate/src/unitary_synthesis.rs

```diff
@@ -0,0 +1,1101 @@
+            coupling_edges,
+            target,
+        )?;
+        match decomposer_item.decomposer {
+            DecomposerType::TwoQubitBasisDecomposer(_) => {
+                let synth = synth_su4_sequence(
+                    &unitary,
+                    decomposer_item,
+                    preferred_dir,
+                    approximation_degree,
+                )?;
+                apply_synth_sequence(py, out_dag, out_qargs, &synth)?;
+            }
+            DecomposerType::XXDecomposer(_) => {
+                let synth = synth_su4_dag(
+                    py,
+                    &unitary,
+                    decomposer_item,
+                    preferred_dir,
+                    approximation_degree,
+                )?;
+                apply_synth_dag(py, out_dag, out_qargs, &synth)?;
+            }
+        }
+        return Ok(());
+    }
+
+    let mut synth_errors_sequence = Vec::new();
+    let mut synth_errors_dag = Vec::new();
+    for decomposer in &decomposers {
+        let preferred_dir = preferred_direction(
+            decomposer,
+            ref_qubits,
+            natural_direction,
+            coupling_edges,
+            target,
+        )?;
+        match &decomposer.decomposer {
+            DecomposerType::TwoQubitBasisDecomposer(_) => {
+                let sequence =
+                    synth_su4_sequence(&unitary, decomposer, preferred_dir, approximation_degree)?;
+                let scoring_info = sequence
+                    .gate_sequence
+                    .gates()
+                    .iter()
+                    .map(|(gate, params, qubit_ids)| {
+                        let inst_qubits =
+                            qubit_ids.iter().map(|q| ref_qubits[*q as usize]).collect();
+                        match gate {
+                            Some(gate) => (
+                                gate.name().to_string(),
+                                Some(params.iter().map(|p| Param::Float(*p)).collect()),
+                                inst_qubits,
+                            ),
+                            None => (
+                                sequence
+                                    .decomp_gate
+                                    .operation
+                                    .standard_gate()
+                                    .name()
+                                    .to_string(),
+                                Some(params.iter().map(|p| Param::Float(*p)).collect()),
+                                inst_qubits,
+                            ),
+                        }
+                    })
+                    .collect::<Vec<(
+                        String,
+                        Option<SmallVec<[Param; 3]>>,
+                        SmallVec<[PhysicalQubit; 2]>,
+                    )>>()
+                    .into_iter();
+                synth_errors_sequence.push((sequence, synth_error(py, scoring_info, target)));
```

Comment:
Removing this collection here might have a positive impact in performance. Just perform the call to `synth_erorr` before pushing the sequence into the `Vec`.
Try this:
```suggestion
                    });
                    let synth_error_from_target = synth_error(py, scoring_info, target);
                synth_errors_sequence.push((sequence, synth_error_from_target));
```

It should hopefully perform better.

Reply:
Thanks for the suggestion!